### PR TITLE
Continuous time smoothing in LinearSDE #343

### PR DESCRIPTION
**Pull request by jzenn**
I hope I've done everything correctly for my first PR here :)

I have three comments

1. I use `np.linalg.inv` to compute the inverse of a matrix. I adapted this from the discrete case. Wouldn't one prefer using a Cholesky decomposition instead to avoid `np.linalg.inv`?
2. For testing the continuous-time smoothing I needed to adapt the `atol` in `np.testing.assert_allclose` to `1e-9` in order for the test to pass. I think this is due to `scipy.integrate.solve_ivp`'s behaviour.
3. I am not quite sure when to include a type hint for the return type for a method.

### CODE REVIEW

**Review by pnkraemer — COMMENTED** [review]
Thanks a lot for this PR! It fits neatly into the rest of the code :) 

Please see my specific comments for detailed feedback. All of them are super minor, and quite subjective as well. 

Regarding your questions:
1. Yes, let's replace it. See my comment below.
2. That is fine. `1e-9` is still sufficiently strict.
3. Since the surrounding functions do not use types (yet), it is fine to not include them in this PR. I will take care of this at some point in the future. 

Another point: there is currently a lack of test coverage. I suppose it is because of this ValueError that is thrown when no `dt` is provided. Would you be so kind and catch that in a test? I suppose the only thing that needs to be added is a test in `TestLinearSDE` that reads along the lines of 
```
def test_backward_realization_value_error_caught(self, normal_rv1, normal_rv2):
    with pytest.raises(ValueError):
            self.transition.backward_realization(
                some_normal_rv1.sample(), some_normal_rv2, t=0.0,
            )
```
and since `backward_realization` is tested, `backward_rv` is automatically tested, too. While at it, could you maybe paste the same test into a function that catches the same error for `forward_realization`? Thanks a lot! 

Thanks Johannes!

File: src/probnum/statespace/sde.py [hunk 1]
```diff
@@ -248,6 +298,39 @@ def f(t, y):
...
+
+            # Apply iteration
+            F = self.driftmatfun(t)
+            u = self.forcevecfun(t)
+            L = self.dispmatfun(t)
+
+            LL = _diffusion * L @ L.T
+            LL_inv_cov = LL @ np.linalg.inv(mde_forward_sol_cov(t))
+
+            new_mean = F @ mean + LL_inv_cov @ (mean - mde_forward_sol_mean(t)) + u
```
[the remark this answers is not in the document]
Comment (jzenn) [reply src/probnum/statespace/sde.py:322]: Looking more closely at the differential equations I don't see how one could use `solve`, since e.g. the smoothing-covariance matrix is some kind of Lyapunov equation. Did I miss something?

File: src/probnum/statespace/sde.py [hunk 2]
```diff
@@ -248,6 +298,39 @@ def f(t, y):
 
         return f, y0
 
+    def _setup_vectorized_mde_backward(self, initrv_obtained, _diffusion=1.0):
```
[the remark this answers is not in the document]
Comment (jzenn) [reply src/probnum/statespace/sde.py:301]: Yes, you're right.

File: src/probnum/statespace/sde.py [hunk 3]
```diff
@@ -217,6 +231,43 @@ def _solve_mde_forward(self, rv, t, dt, _diffusion=1.0):
...
+        sol = scipy.integrate.solve_ivp(
+            mde,
+            (t + dt, t),
+            y0,
+            method=self.mde_solver,
+            atol=self.mde_atol,
+            rtol=self.mde_rtol,
+            # Use forward solution for mean and covariance in scipy's ivp
+            args=(mde_forward_sol_mean, mde_forward_sol_cov),
+            # Dense output for lambda-expression
```
[the remark this answers is not in the document]
Comment (jzenn) [reply src/probnum/statespace/sde.py:254]: I can do that if you're feeling nervous about it :)

File: src/probnum/statespace/sde.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jzenn) [reply src/probnum/statespace/sde.py:322]: - Good idea.
- I use `solve` as you suggested

**Comment by jzenn** [comment]
> Another point: there is currently a lack of test coverage. I suppose it is because of this ValueError that is thrown when no `dt` is provided. Would you be so kind and catch that in a test? I suppose the only thing that needs to be added is a test in `TestLinearSDE` that reads along the lines of
> 
> ```
> def test_backward_realization_value_error_caught(self, normal_rv1, normal_rv2):
>     with pytest.raises(ValueError):
>             self.transition.backward_realization(
>                 some_normal_rv1.sample(), some_normal_rv2, t=0.0,
>             )
> ```
> 
> and since `backward_realization` is tested, `backward_rv` is automatically tested, too. While at it, could you maybe paste the same test into a function that catches the same error for `forward_realization`? Thanks a lot!

Since `TestLTISDE` inherits from `TestLinearSDE` I would need to add the two test to this class to (due to `matrix_fraction_decomposition` failing with `dt=None`). Is that alright or did I miss something?

**Comment by pnkraemer** [comment]
Hmmmm seems like this issue of `dt=None` is not caught in `LTISDE` methods at all... Ideally, these functions should be in `TestLinearSDE` and then automatically test on instances of `LTISDE` (which also require `dt != None`).

In any case, for this PR, I'd say to keep things consistent. So either remove the `if dt is None` checks from `LinearSDE` (we can write an issue and someone here can take care of this some other time) or write the general tests and add the checks to `LTISDE`. It is your call :) What do you think?

**Comment by jzenn** [comment]
I can just add the check for `dt != None` to `LTISDE` that is no problem. Thank you!

**Comment by jzenn** [comment]
I also adapt the corresponding parts of `Matern` and `IBM` which would also be affected from the test.

**Comment by jzenn** [comment]
In my opinion the test-failures seem to be unrelated to any changes I made. I just noticed you're experiencing the same in #365.

**Comment by pnkraemer** [comment]
Yes, the test failures are not your fault...

**Comment by pnkraemer** [comment]
Try updating the newest master branch, the CI issues should be fixed. If not, please let me know :)

**Review by pnkraemer — APPROVED** [review]
Looks good, thanks for your efforts (and for your patience)! 
One optional piece of feedback:
If you could put a one-liner docstring into the value error caught tests, that would be great. This is entirely optional, though, and we can merge either way. 

If you would like to continue, please let me know. Thanks a lot for you contribution either way!

**Comment by jzenn** [comment]
> If you could put a one-liner docstring into the value error caught tests, that would be great. This is entirely optional, though, and we can merge either way.

I can do that.

**Comment by pnkraemer** [comment]
Great, thanks again! Everything is fine from my side, so let me know when you want me to merge this for you (I dont think you can?)!

**Comment by jzenn** [comment]
From my side you can merge. I thought closing this PR would merge automatically :)

**Comment by jzenn** [comment]
Sorry for the inconvenience this has caused, #371 adds (only) the continuous-time smoother for LinearSDE.
